v1/plugins/logs/encoder.go

```diff
@@ -67,22 +74,44 @@ func (enc *chunkEncoder) Write(event EventV1) (result [][]byte, err error) {
 // WriteBytes attempts to write a serialized event to the current chunk.
 // If the upload limit is reached the chunk is closed and a result is returned.
 // The incoming event that didn't fit is added to the next chunk.
-func (enc *chunkEncoder) WriteBytes(bs []byte) (result [][]byte, err error) {
+func (enc *chunkEncoder) WriteBytes(bs []byte) ([][]byte, error) {
 	if len(bs) == 0 {
 		return nil, nil
-	} else if int64(len(bs)+2) > enc.limit {
+	}
+
+	// Compress the incoming event by itself in order to verify its final size
+	incomingEventBuf := new(bytes.Buffer)
+	w := gzip.NewWriter(incomingEventBuf)
+	if _, err := w.Write(bs); err != nil {
+		return nil, err
+	}
+	if err := w.Flush(); err != nil {
+		return nil, err
+	}
+
+	// if the compressed incoming event is bigger than the upload size limit reject it
+	if int64(incomingEventBuf.Len()+2) > enc.limit {
 		if enc.metrics != nil {
 			enc.metrics.Counter(encLogExUploadSizeLimitCounterName).Incr()
 		}
-		return nil, fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
-			int64(len(bs)+2), enc.limit)
+		return nil, fmt.Errorf("received a decision event with size %d that exceeds the upload_size_limit_bytes %d",
+			int64(incomingEventBuf.Len()), enc.limit)
 	}
 
-	if int64(len(bs)+enc.bytesWritten+1) > enc.softLimit {
+	if err := enc.w.Flush(); err != nil {
+		return nil, err
+	}
+
+	// If adding the compressed incoming event to the current compressed buffer exceeds the limit,
+	// close the current chunk and reset it so the incoming event can be written into the next chunk.
+	// Note that the soft limit enforces the final compressed size.
+	var result [][]byte
+	if int64(incomingEventBuf.Len()+enc.buf.Len()+1) > enc.softLimit {
```

Comment:
I know this isn't your invention, but this `1` does a lot of heavy lifting. Depending on what's already written to `enc.buf`, we'll either need to write an extra `[`, `,`, `]`, or a combination thereof. Assuming a chunk will contain >2 events most of the time, `1` (`,`) is mostly right 😄.

Reply:
Considering things like this, and the various gzip headers and trailers that will be added, and the fact that the compressed forms of `[,]` might not be exactly one byte each, it doesn't look like we'll ever firmly respect the configured upload limit. But given that the soft-limit adds even more variability, I suppose that's ok.